# 第3次作业 （共4题）

## 第1题（10分）


在这一题中，你将根据一个[Polybius加密算法](https://ctf-wiki.org/crypto/classical/polyalphabetic/)，实现对一段文本的加密和解密。约定文本仅由大写字母书写而成。


Polybius密码又称为棋盘密码，加密过程是将给定的明文根据密码表加密为两两组合的数字，密码表由一个$5 \times 5$的矩阵表示。下面给出了一个常用密码表。

```
|   	| 1 	| 2 	| 3 	| 4   	| 5 	|
|---	|---	|---	|---	|-----	|---	|
| 1 	| A 	| B 	| C 	| D   	| E 	|
| 2 	| F 	| G 	| H 	| I/J 	| K 	|
| 3 	| L 	| M 	| N 	| O   	| P 	|
| 4 	| Q 	| R 	| S 	| T   	| U 	|
| 5 	| V 	| W 	| X 	| Y   	| Z 	|
```

记密码表矩阵为$A$，其第$i$行第$j$列（从1开始计数）的元素记为$A_{ij}$。该密码表的含义为，将文本中所有出现的$A_{ij}$字母，替换为数字$ij$。

比如，将文本中所有出现的'L'字母，替换为数字$31$。由于英文字母有26个，比矩阵的元素个数多出一个，因此约定将I、J两个字母均视为字母I。

如果待加密文本为`HELLO`，那么根据上述密码表进行加密后，结果应为`2315313134`。如果再对这个结果进行解密，应该得到`HELLO`。

请在下方的代码单元格中，基于字典推导式，完成如下两个函数的定义。

+ `encrypt(text, codebook)`：其中，text为待加密文本（一个字符串），codebook为密码表（一个列表），函数需返回加密后的文本（一个字符串）。
+ `decrypt(text, codebook)`：其中，text为待解密文本（一个字符串），codebook为密码表（一个列表），函数需返回解密后的文本（一个字符串）。

密码表`codebook`是一个嵌套列表。其第$i$个元素（$i$从0计数），`codebook[i]`，是一个列表，对应$A$矩阵的第$i+1$行的所有元素。该列表中的第$j$个元素（$j$从0计数），`codebook[i][j]`，对应矩阵元素$A_{(i+1)(j+1)}$。`codebook[i][j]`本身是一个字符，表示一个英文大写字母。比如，示例中的密码表对应的`codebook`如下。
```
[['A', 'B', 'C', 'D', 'E'],
 ['F', 'G', 'H', 'I', 'K'],
 ['L', 'M', 'N', 'O', 'P'],
 ['Q', 'R', 'S', 'T', 'U'],
 ['V', 'W', 'X', 'Y', 'Z']]
```
注意到`codebook`中不包含字母`J`，因为根据约定，文本中所有出现的`J`都会被视为`I`。

加密函数解密函数各占一半分数。请使用你定义的函数跑通这一题的后续代码，并展示输出。

In [9]:
# 请不要改动这里的代码
codebook = [['A', 'B', 'C', 'D', 'E'],
 ['F', 'G', 'H', 'I', 'K'],
 ['L', 'M', 'N', 'O', 'P'],
 ['Q', 'R', 'S', 'T', 'U'],
 ['V', 'W', 'X', 'Y', 'Z']]

In [10]:
# 你的作答
def encrypt(text, codebook):
    str_i = ""
    str_j = ""
    result1 = ""
    
    for ch in text:
        for i in range(5):
            for j in range(5):
                if ch == codebook[i][j]:
                    str_i = str(i + 1)
                    str_j = str(j + 1)
                    result1 += str_i + str_j
                    
    return result1

def decrypt(text, codebook):
    nums_code = len(text)
    num_i = 0
    num_j = 0
    result2 = ""

    for num in range(0,nums_code,2):
        num_i = int(text[num]) - 1
        num_j = int(text[num + 1]) - 1
        result2 += codebook[num_i][num_j]
    
    return result2

In [11]:
# 不要改动这里的代码
text1 = 'HELLO'
print(encrypt(text1, codebook))
print(decrypt(encrypt(text1, codebook), codebook))



2315313134
HELLO


注意到`codebook`中的字母顺序即便随意打乱，编解码过程应该也能跑的通。打乱字母顺序相当于更换用来加密和解密的密码本。请试试利用下面的新密码本，解码给定的密文。

In [12]:
# 请不要改动下方的代码
import random
from string import ascii_uppercase

random.seed(101)

letters = list(ascii_uppercase.replace('J', ''))
random.shuffle(letters)
letters = ''.join(letters)

codebook = []
for i in range(5):
    row = []
    for j in range(5):
        idx = i*5 + j
        row.append(letters[idx])
    codebook.append(row)
    
    
text2 = '1222312322534214445522245323514355421433'
decrypt(text2, codebook)

'LIFEISHORTIUSEPYTHON'

## 第2题（20分）


博物馆名画失窃案（进阶逻辑版）

名画失窃后，警方锁定了 A, B, C, D, E 五名嫌疑人。经过深度调查，警方掌握了以下背景：
* 5 人中恰好有 **2 名小偷**、**1 名帮凶**和 **2 名清白者**。
* 诚实准则：
    1.  **清白者**：双目如炬，2 条证词**全为真话**。
    2.  **小偷**：老谋深算，2 条证词**恰好 1 真 1 假**。
    3.  **帮凶**：由于极度心虚，2 条证词**全为假话**。

审问记录如下：
+ **A 说**：1.  C 是小偷。2.  如果 B 是小偷，那么我（A）是清白的。
+ **B 说**：1.  E 是小偷。2.  D 是帮凶。
+ **C 说**：1.  A 是帮凶。2.  如果 B 是小偷，那么 E 也是小偷。
+ **D 说**：1.  E 是小偷。2.  B 和 E 都是小偷。
+ **E 说**：1.  C 是帮凶。2.  A 绝不是清白的人（即 A 是小偷或帮凶）。

**请问：谁是小偷？谁是帮凶？**

**说明**：条件句“如果 P，那么 Q”仅在 P 为真且 Q 为假时为“假”；若前提 P 为假，则该证词视为“真”。

In [19]:
from itertools import combinations

suspects = ['A', 'B', 'C', 'D', 'E']

testimonies = {
    "A": [("C", "thief"), ("if", ("B", "thief"), ("A", "innocent"))],
    "B": [("E", "thief"), ("D", "accomplice")],
    "C": [("A", "accomplice"), ("if", ("B", "thief"), ("E", "thief"))],
    "D": [("E", "thief"), ("and", ("B", "thief"), ("E", "thief"))],
    "E": [("C", "accomplice"), ("not", ("A", "innocent"))],
}

def evaluate(prop, roles):
    if isinstance(prop, tuple):
       
        if prop[0] == "if":
            _, antecedent, consequent = prop
            return (not evaluate(antecedent, roles)) or evaluate(consequent, roles)
        elif prop[0] == "and": 
            _, clause1, clause2 = prop
            return evaluate(clause1, roles) and evaluate(clause2, roles)
        elif prop[0] == "not": 
            _, clause = prop
            return not evaluate(clause, roles)
        elif len(prop) == 2:
            person, role = prop
            return roles[person] == role
    return False

def check_truth_pattern(truths, role):
    if role == "innocent":
        return truths == (True, True)
    elif role == "accomplice":
        return truths == (False, False)
    elif role == "thief":
        return truths in [(True, False), (False, True)]
    return False

def solve():
    all_people = suspects
    
    for thieves in combinations(all_people, 2):
        rest = [p for p in all_people if p not in thieves]
        for accomplice in rest:
            innocents = [p for p in rest if p != accomplice]
            
            roles = {}
            for p in all_people:
                if p in thieves:
                    roles[p] = "thief"
                elif p == accomplice:
                    roles[p] = "accomplice"
                else:
                    roles[p] = "innocent"
            
            valid = True
            for person in all_people:
                stmt1, stmt2 = testimonies[person]
                truth1 = evaluate(stmt1, roles)
                truth2 = evaluate(stmt2, roles)
                truths = (truth1, truth2)
                
                if not check_truth_pattern(truths, roles[person]):
                    valid = False
                    break
            
            if valid:
                for p in all_people:
                    print(f"  {p}: {roles[p]}")
                return
    
    print("未找到符合所有条件的解")

if __name__ == "__main__":
    solve()

  A: accomplice
  B: thief
  C: innocent
  D: innocent
  E: thief


## 第3题（30分）

这道题将带你深入理解Python中浮点数表示的底层原理。让我们从数学常数$\pi$开始。`math`库中定义了$\pi$的值，即`math.pi`。我们除了可以直接查看它的取值，还可以通过`struct.pack()`函数查看它在Python中的二进制表示。

In [20]:
# 不要改动下方代码
import struct
import math

print("math.pi：", math.pi)
print("math.pi的字节码表示：", struct.pack('<d', math.pi))

math.pi： 3.141592653589793
math.pi的字节码表示： b'\x18-DT\xfb!\t@'


如何解读上述结果？首先要明确，`math.pi`是以64位浮点数的形式存储的。所谓64位，是指64个比特（bit），每个比特是一个二进制数字，即0或1。那么，`math.pi`到底对应着哪64个二进制数字的组合呢？我们通过`struct.pack`函数获得的结果，实际上是这64个二进制数字对应的“字节码（byte code）”表示。因此，我们需要先理解字节码表示与二进制表示之间的关联。

+ 1个字节相当于8个比特，而8个比特，本质上是8个二进制数的排列组合，一共有$2^8=256$种组合。比如，某个字节存储的信息可以为`00111111`，属于这256种组合中的一个。
+ 因此，1个字节所表示的信息，通过二进制到十进制的转换，也可以由一个0到255之间的十进制整数来表示。比如，`00111111`作为1个字节，所表示的信息，可等价的由十进制整数63表示。可以通过代码`int('00111111', base=2)`来完成这一转换过程。
+ 相应的，给定一个十进制整数`x`，我们可以通过代码`format(x, '08b')`得到它对应的8位二进制表示。比如，`format(63, '08b')`返回字符串`'00111111'`。这段代码中，`b`表示我们希望将十进制的`x`转换为它的二进制表示，并以字符串的形式返回。而`08b`表示让这个二进制字符串至少占8个字符，不足8个字符时在前面补0。
+ 因此，一个64位浮点数对应8个字节的信息，可等价的由8个0到255之间的十进制整数来表示。
+ 对于任意一个0到255之间的十进制整数，可以基于字节码的编码规则，将其表示为若干字符。该规则如下。
  - [ASCII字符集](https://www.ascii-code.com/)中一共定义了127个字符，每个字符对应一个0到126的十进制整数。比如，ASCII字符`?`对应十进制整数63。
  - 因此，当1个字节对应的十进制整数取值在0到127之间时，我们使用这个整数对应的ASCII字符来表示该字节。比如，字节`00111111`对应的十进制整数为63，而十进制整数63对应的ASCII字符为`?`。因此，字节`00111111`的字节码表示为`b'?'`。其中，前缀`b`表示后面的字符串应该被解读为字节码而非纯文本。可以通过函数`ord`和`chr`分别获得一个ASCII字符的十进制整数编码和一个十进制整数对应的ASCII字符。
  - 当要表示的字节对应的整数在127到255之间时，我们将该整数转换为一个16进制整数。比如，字节`11011011`对应的十进制整数为219，转换为16进制整数后得到'db'。因此，字节`11011011`的字节码表示为`b'\xdb'`。其中，前缀`\x`表示后面的两个字符应该被整体地解读为一个16进制整数，而非单独的两个ASCII字符。可以通过代码`int('db', base=16)`完成从16进制到十进制的转换，代码`hex(219)`完成从十进制到16进制的转换。
+ 基于以上的编码规则，并遵循“小端序”（[little-endian](https://www.geeksforgeeks.org/little-and-big-endian-mystery/)）的字节存储顺序，我们需要从右往左，逐字节理解字节码`b'\x18-DT\xfb!\t@'`中的信息。具体做法如下：
    1. 第1个字节的字节码是`b'@'`，对应ASCII字符`@`，相当于十进制整数64，因此存储的8位比特信息为`01000000`；
    2. 第2个字节的字节码是`b'\t'`，对应ASCII字符`\t`（制表符），相当于十进制整数9，因此存储的8位比特信息为`00001001`；
    3. 第3个字节的字节码是`b'!'`，对应ASCII字符`!`，相当于十进制整数33，因此存储的8位比特信息为`00100001`；
    4. 第4个字节的字节码是`b'\xfb'`，对应16进制整数`fb`，相当于十进制整数为`251`，因此存储的8位比特信息为`11111011`；
    5. 第5个字节的字节码是 ...
    6. 第6个字节的字节码是 ...
    7. 第7个字节的字节码是 ...
    8. 第8个字节的字节码是 ...
+ 最后，将所有信息汇总，`math.pi`的字节码表示为`b'\x18-DT\xfb!\t@'`，对应的二进制表示为`01000000 00001001 00100001 11111011 ... ... ... ...`。

### 第一问（10分）

请根据你对上述的理解，在上方的Markdown单元格中，填充内容中的缺失信息（所有用"..."标记的地方），同时按照要求补全下方代码，进一步验证你的理解。

In [21]:
# 我们可以使用下面的代码，来验证我们的理解。
def bytes2bits(byte_data):
    # 将字节码转换为二进制字符串 (注意小端序：从后往前取字节)
    binary_str = ""
    # 遵循小端序，使用 reversed(byte_data) 遍历字节，并用 format(x, '08b') 转换
    for b in reversed(byte_data):
        # 在 Python 3 中，当你遍历一个原始字节码对象（bytes 类型）时，
        # 循环变量等于每个字节对应的 10 进制整数（取值范围在 0 到 255 之间）
        # 请根据这一事实，补全下方代码
        # ---------------------------------
        two_b = format(b,'08b')
        str_b = str(two_b)
        binary_str += str_b # 请替换这个空字符串
        # ---------------------------------
    return binary_str

raw_bytes = b'\x18-DT\xfb!\t@'
bits = bytes2bits(raw_bytes)
str_len = len(bits)
print(f"数学常数`math.pi`的字节码表示为 {raw_bytes} ，对应的二进制表示为：", ' '.join([bits[i:i+8] for i in range(0, str_len, 8)]))

数学常数`math.pi`的字节码表示为 b'\x18-DT\xfb!\t@' ，对应的二进制表示为： 01000000 00001001 00100001 11111011 01010100 01000100 00101101 00011000


知道了`math.pi`的底层二进制表示后，接下来自然要问：为什么这个二进制数会对应着`math.pi`这个浮点数呢？这就引出了IEEE 754标准。为了更方便地理解，我们先回顾二进制计数法的基本原理。


### 2.1. 二进制计数法

对于十进制整数而言，每一位代表 10 的幂次：$\underbrace{1}_{10^3}\underbrace{2}_{10^2}\underbrace{3}_{10^1}\underbrace{4}_{10^0} = 1\times 1000 + 2\times 100 + 3\times 10 + 4\times 1$。

对于二进制整数而言，每一位只能是 0 或 1，在换算为十进制时，每一位代表 2 的幂次：$(\underbrace{1}_{2^3}\underbrace{1}_{2^2}\underbrace{0}_{2^1}\underbrace{1}_{2^0})_{2} = 1 \times 8 + 1 \times 4 + 0 \times 2 + 1 \times 1 = (13)_{10}$。注意，当我们想明确一个数字的进制时，我们采取“$(\text{数字})_{\text{进制}}$”的写法。

对于十进制小数而言，小数点右边的每一位代表 10 的负幂次：$1\ 2\ .\ 3\ 4 = 1\times10^1 + 2\times10^0 + 3\times10^{-1} + 4\times10^{-2}$。

对于二进制小数而言，在换算为十进制时，小数点右边的每一位代表 2 的负幂次，依次是 $2^{-1}, 2^{-2}, 2^{-3}\ldots$，即 $\frac{1}{2}, \frac{1}{4}, \frac{1}{8}\ldots$。比如：

$$(1\ 1\ 0\ 0\ .\ 1\ 0\ 0\ 1)_{2} = 1\times2^3 + 1\times2^2 + 0\times2^1 + 0\times2^0 + 1\times2^{-1} + 0\times2^{-2} + 0\times2^{-3} + 1\times2^{-4}$$

$$= 8 + 4 + 0 + 0 + 0.5 + 0 + 0 + 0.0625 = (12.5625)_{10}。$$

注意，换算过程中涉及到的运算均为十进制算术。

更一般的，设一个二进制小数为 $b_n b_{n-1} \cdots b_0 . b_{-1} b_{-2} \cdots b_{-m}$。对于任意位置$i$，有$b_i \in \{0,1\}$。约定按以下公式计算其十进制值$x$：
$$(b_n b_{n-1} \cdots b_0 . b_{-1} b_{-2} \cdots b_{-m})_{2} = \sum_{i=-m}^{n} b_i \times 2^i = (x)_{10}。$$

### 2.2. 二进制的科学记数法

我们可以把任意十进制数写成科学计数法的形式：

$$\pm\ d_0.d_{-1} d_{-2} d_{-3} \cdots \times 10^n \quad (1 \leq d_{i} < 10)，$$

即小数点左侧只保留一位非零数字。右边乘以 $10^n$ 的本质是**移动小数点**：

- $\times 10^3$：小数点右移 3 位（数变大）
- $\times 10^{-2}$：小数点左移 2 位（数变小）

也就是说，乘幂等价于移位。这是因为每一位的权重是10的幂次，故而乘幂相当于提高或降低每一位的权重的指数，即移动小数点。考虑十进制数$1.234 = 1\times10^0 + 2\times10^{-1} + 3\times10^{-2} + 4\times10^{-3}$。两边同乘 $10^3$：
$$1.234 \times 10^3 = (1\times10^0 + 2\times10^{-1} + 3\times10^{-2} + 4\times10^{-3}) \times 10^3 = 1\times10^3 + 2\times10^2 + 3\times10^1 + 4\times10^0 = 1234。$$
即每一项权重的指数都 **+3**，视觉效果就是小数点向右移动了 3 位。

我们可以类比地定义二进制的科学计数法。任意二进制数可以被**规格化**（normalize）为：
$$\pm\ 1.b_{-1} b_{-2} b_{-3} \cdots \times 2^n，$$
其中 $n$ 是**十进制整数**，表示**移动的位数**，不是二进制数。
+ 与十进制的科学计数法类似，规格化后的二进制小数，小数点左侧只保留一位非零数字。不同于十进制非零位有1-9一共9种数字，二进制的非零位只能可能是 1。所以规格化后，小数点前**永远是 1**。
+ 正如十进制中乘以$10^n$相当于移动十进制数的小数点，上式中，乘以 $2^n$ 相当于移动二进制数的小数点，机制同十进制时一样：$n>0$相当于将小数点向右移动$n$位，$n<0$相当于将小数点向左移动$n$位。比如，$1100.1$ 可以被规格化为 $1.1001 \times 2^3$。我们可以如下理解这一过程。

二进制数$1.1001$换算为十进制为：$(1.1001)_{2} = 1\times2^0 + 1\times2^{-1} + 0\times2^{-2} + 0\times2^{-3} + 1\times2^{-4}$。两边同乘 $2^3$（这里的 3 是十进制数）：

$$(1.1001)_{2} \times 2^3 = (1\times2^0 + 1\times2^{-1} + 0\times2^{-2} + 0\times2^{-3} + 1\times2^{-4}) \times 2^3$$

$$= 1\times2^3 + 1\times2^2 + 0\times2^1 + 0\times2^0 + 1\times2^{-1} = (1100.1)_{2}$$

也就是说，乘以$2^3$相当于把$1.1001$的小数点向右移动3位。注意到：一方面，$1.1001 \times 2^3$相当于十进制的乘法运算$ 1.5625 \times 8 = 12.5$（二进制的$1.1001$等于十进制的$1.5625$）；另一方面，通过移位得到的 $1100.1$换算为十进制为 $8+4+0.5 = 12.5$。两种运算方式的结果相等，符合我们对于乘法运算的理解：$1.1001 \times 2^3$相当于把$1.1001$代表的数加8遍。这正如二进制数$(10)_2=(1.0)_2 \times 2^1$等于$(1.0)_2+(1.0)_2$一样（即将二进制数$1$加两遍得到二进制数$10$，也就是十进制的$2$）。



### 2.3. IEEE 754 双精度浮点数

既然任意二进制数都能写成 $\pm\ 1.b_{-1} b_{-2} b_{-3} \cdots \times 2^n$ 的形式，且小数点左侧永远是 1，因此计算机只需存储三份信息：正负号，指数$n$（十进制整数），以及小数点后的尾数$b_{-1} b_{-2} b_{-3} \cdots$（二进制整数）。IEEE 754 双精度浮点数的标准中，约定一共使用64位（即64 个比特），按如下方式划分字段，来存储这三份信息。

| 需要存的信息 | 对应字段 | 位数（即使用几个比特存该信息） |
|------------|---------|------|
| 正负号 | 符号位（Sign） | 1 位, 0 表示正数, 1 表示负数 |
| 指数 $n$（十进制整数） | 指数位（Exponent） | 11 位, 加偏移量1023后以二进制形式存储 |
| 小数点后的 $b_{-1} b_{-2} b_{-3} \cdots$ | 尾数位（Mantissa / Fraction） | 52 位 |

说明：
+ 十进制的指数 $n$ 以 11 位的二进制形式存储。原则上，指数$n$可正可负，标准中约定它最小为 $-1022$，最大为 $1023$。为了避免在这11位中单独留出一位来存储指数的符号，标准约定将指数$n$加上偏移量 $1023$，确保结果一定为正，再以二进制形式存储。例如，指数 $n=3$ 以十进制数 $3+1023=1026$ 存储在指数位中，对应的二进制数为 $10000000010$。
+ 由于指数位可以存储$2^{11}=2048$种二进制数的组合，而标准约定的指数范围$ -1022 \le n \le 1023$只占据了其中的$1023-(-1022)+1=2046$种组合，我们还有2组特殊的二进制组合没有使用，分别为11个0或者11个1。对它们的含义和触发场景做如下约定。

| 指数位 | 尾数位 |  特殊值 | 触发场景 |
|--------|--------|--------|---------|
| 全0 | 全0 | $\pm 0$ | 零值 |
| 全0 | 非全0 | 非规格化数 | 极小值，接近0 |
| 全1 | 全0 | $\pm \infty$ | 溢出、除以零 |
| 全1 | 非全0 | NaN (Not a Number) | $0/0$、$\infty-\infty$、$\sqrt{-1}$ |


最后，以十进制数$12.5$为例，展示将其存储为双精度浮点数的完整流程。第一步，转二进制后，得到$1100.1$。第二步，规格化后，得到$1.1001 \times 2^3$。第三步，拆成三部分，分别为：符号位（0），指数位（3+1023=1026），尾数位（1001）。指数位转换为二进制数为$10000000010$。尾数位本就为二进制，但是位数不足52位，需要在末尾补齐0，得到$1001000...0$。最终的64位表示如下：

```
0 | 10000000010 | 1001000000000000000000000000000000000000000000000000
↑   ────┬──────   ───────────────────────┬────────────────────────────
符号  指数(11位)                      尾数(52位)
```

总结：当我们想存储一个十进制小数时，要先将其换算为二进制小数并规格化，再根据IEEE 754 双精度浮点数的标准，将其严格塞进 64 个比特里——符号位管正负，指数位管小数点的位置，尾数位管精度。

### 第二问（10分）

请根据你对于上述内容的理解，补全下方函数`decode_double_precision`，将一个IEEE 754 双精度浮点数的二进制表示转换为十进制数，并查看对于下方给定数字的转换结果。

In [30]:
# 你的作答
def decode_double_precision(ieee754_binary_str):
    """
    将一个64位的IEEE 754 双精度浮点数的二进制表示转换为十进制数
    参数：
      ieee754_binary_str：一个IEEE 754 双精度浮点数的二进制表示，字符串格式，长度64
      比如："0100000000001001001000011111101101010100010001000010110100011000"
    返回值：
      该IEEE 754 双精度浮点数的十进制表示，Python原生浮点数格式
      比如：3.141592653589793
    """
    result_str = ""
    result_double = 0
    power_ten = 0
    power_two_str = ""
    power_two = 0
    num_ten = 0
    num_two_str = "1."
    num_two = 0
    num_str = ""
    num = 0

    if ieee754_binary_str[0] == '1':
        result_str = result_str + "-"

    for i in range(1,12):
        power_two_str = power_two_str + ieee754_binary_str[i]
    power_ten = int(power_two_str,base = 2) - 1023

    for j in range(12,64):
        num_two_str = num_two_str + ieee754_binary_str[j]
    for k in range(52):
        if num_two_str[k+2] == '1':
            num_ten += 2**(-(k+1))

    num = 1.0 + num_ten
    if power_ten >= 0:
        for k in range(power_ten):
            num = num * 2
    else:
        for k in range(-power_ten):
            num = num / 2
    
    num_str = str(num)
    result_str = result_str + num_str
    
    result_double = float(result_str)
    return result_double

In [31]:
# 不要改动下方的代码
binary_str = '0100000000000101101111110000101010001011000101000101011101101001'
print(decode_double_precision(binary_str))

2.718281828459045


观察下方代码的运行结果，思考为什么会出现这个结果。

In [2]:
0.1 + 0.1 + 0.1 == 0.3

False

### 第三问（10分）

根据IEEE 754 双精度浮点数的定义，`0.1`应该如何存储？通过双精度浮点数，能将`0.1`精确的存储下来么？这对于浮点数运算`0.1+0.1+0.1`的精度有何影响？将你的作答写在下方的Markdown单元格中。


1、符号位：0 指数位：01111111011（二进制）= 1019（十进制），真实指数 = 1019 - 1023 = -4 尾数位：1001100110011001100110011001100110011001100110011010 
2、不能 0.1在二进制中是无限循环小数
3、0.1的储存值其实并不是0.1，而是一个和0.1十分相近的数。三次累加后误差会被放大。

### 附录：

将十进制纯小数转换为二进制，采用的是“乘 2 取整，顺序排列”法：
1.  将十进制小数乘以 **2**。
2.  取出结果的**整数部分**（0 或 1），作为二进制小数该位的数字。
3.  取出结果的**小数部分**，重复步骤 1，直到小数部分为 0 或达到精度限制。

以十进制纯小数 `0.625`为例， 带入上述算法进行计算：
* $0.625 \times 2 = 1.25$ $\rightarrow$ 取整数部分 **1**
* $0.25 \times 2 = 0.5$ $\rightarrow$ 取整数部分 **0**
* $0.5 \times 2 = 1.0$ $\rightarrow$ 取整数部分 **1**

最终得到 $(0.625)_{10} = (0.101)_2$。

下面，我将从二进制小数的数学定义出发，利用代数推导来证明十进制纯小数通过“乘 2 取整”法可以得到正确的二进制表示。

#### A.1. 二进制小数的数学定义

任何一个十进制纯小数 $x$（满足 $0 \le x < 1$），在二进制下的表示可以写作：
$$(x)_{10} = (0.b_1b_2b_3...b_n...)_2$$
这意味着 $x$ 可以表示为以 $2$ 的负幂次为权重的累加和：
$$x = b_1 \cdot 2^{-1} + b_2 \cdot 2^{-2} + b_3 \cdot 2^{-3} + \dots + b_i \cdot 2^{-i} + \dots$$
其中，每一个系数 $b_i$ 只能是 **0** 或 **1**。

#### A.2. 数学证明过程：逐步剥离系数

我们要证明的是：通过将 $x$ 乘以 $2$ 并提取整数部分，可以依次得到 $b_1, b_2, b_3, \dots$。

+ 第一步：提取第一位系数 $b_1$。将等式两边同时乘以 **2**：
  $$2x = 2 \cdot (b_1 \cdot 2^{-1} + b_2 \cdot 2^{-2} + b_3 \cdot 2^{-3} + \dots)$$
  $$2x = b_1 + (b_2 \cdot 2^{-1} + b_3 \cdot 2^{-2} + \dots)$$
  此时，等式右边分为两部分：
  1.  **整数部分**：$b_1$。
  2.  **小数部分**：$(b_2 \cdot 2^{-1} + b_3 \cdot 2^{-2} + \dots)$。

+ **关键证明点**：为什么小数部分一定小于 1？因为二进制小数的最大值（当所有 $b_i$ 都为 1 时）是一个无穷递缩等比数列：
  $$\sum_{i=2}^{\infty} 2^{-(i-1)} = \frac{1}{2} + \frac{1}{4} + \frac{1}{8} + \dots = 1$$
  实际上，除非从某一位开始全是 1（这在规范化表示中会进位），否则该部分和总是严格小于 1。因此，$2x$ 的**整数部分**必然等于 $b_1$。

+ 第二步：提取第二位系数 $b_2$。令第一步剩下的剩余小数部分为 $x_1$：
  $$x_1 = b_2 \cdot 2^{-1} + b_3 \cdot 2^{-2} + \dots$$
  再次两边乘以 **2**：
  $$2x_1 = b_2 + (b_3 \cdot 2^{-1} + \dots)$$
  同理，$2x_1$ 的整数部分就是 $b_2$。

+ 归纳推理：以此类推，第 $n$ 次“乘 2 取整”得到的整数就是二进制表示中第 $n$ 位的值 $b_n$。

## 第4题（40分）

### 背景故事

某城市正在开展一项大规模体检活动。已知如下信息：

- 在该城市的全部人口中，**每 100 个人里大约有 1 个人**真正患有某种疾病。
- 体检使用的检测试剂有如下特点：
  - 如果一个人**真的患病**，检测结果显示"阳性"（即被查出来）的可能性为 **99%**；
  - 如果一个人**实际上健康**，检测结果却错误地显示"阳性"（即虚惊一场）的可能性为 **5%**。

现在，小明参加了体检，报告显示**阳性**。他非常紧张地问你：

> **"我真的得病了吗？检测结果为阳性，我真正患病的可能性究竟有多大？"**

请你先凭直觉猜一个数字，写在纸上。然后，我们用 Python 模拟整个城市的体检过程，通过统计数据来验证你的直觉。

### 第一问（10分）：模拟一个人的体检过程

编写一段代码，模拟**一个人**的体检过程，并将结果用一个**元组**表示。具体步骤如下：

1. 使用 `random`库的`random()`函数生成一个随机数，判断这个人是否患病（随机数小于 `0.01` 则患病）。
2. 根据是否患病，再用 `random()`函数生成另一个随机数来决定检测结果：
   - 如果患病：随机数小于 `0.99` 则检测为阳性。
   - 如果健康：随机数小于 `0.05` 则检测为阳性。
3. 用元组 `(是否患病, 检测结果)` 记录此人信息，`True` 表示患病/阳性，`False` 表示健康/阴性。
4. 打印这个元组，并用三元表达式将结果翻译成中文输出。比如，元组`(True, False)`应该翻译为`'此人实际状况：患病，检测结果：阴性'`。


In [21]:
# 你的作答
from random import random # 不要删除这行代码
probability = random()

if probability < 0.01: # 患病
    ill_res = True
    test_pro = random()
    if test_pro < 0.99:
        test_res = True
    else:
        test_res = False
    test = "阳性" if test_pro < 0.99 else "阴性"
else: # 健康
    ill_res = False
    test_pro = random()
    if test_pro < 0.05:
        test_res = True
    else:
        test_res = False
    test = "阳性" if test_pro < 0.05 else "阴性"

result_tuple = (ill_res,test_res)
ill = "患病" if ill_res == True else "健康"
print("此人实际状况：",ill,"，检测结果：",test)

此人实际状况： 健康 ，检测结果： 阴性


### 第二问（10分）：模拟整座城市的体检

将第一问的逻辑封装为一个**函数** `simulate_one_person()`，然后模拟一座有 100,000 人的城市。
1. 定义函数 `simulate_one_person()`，不需要任何参数，返回一个元组 `(是否患病, 检测结果)`。
2. 使用**列表推导式**，调用该函数 100,000 次，将所有人的元组收集到列表 `city_data` 中。
3. 打印 `city_data` 的前 10 个元素，确认数据格式正确。


In [30]:
# 你的作答
from random import random

def simulate_one_person():
    probability = random()

    if probability < 0.01: # 患病
        ill_res = True
        test_pro = random()
        if test_pro < 0.99:
            test_res = True
        else:
            test_res = False
    else: # 健康
        ill_res = False
        test_pro = random()
        if test_pro < 0.05:
            test_res = True
        else:
            test_res = False

    result_tuple = (ill_res,test_res)
    return result_tuple

city_data = [simulate_one_person() for i in range(100000)]

for j in range(10):
    print(city_data[j])

(False, False)
(False, False)
(False, False)
(False, False)
(False, False)
(False, False)
(False, False)
(False, False)
(False, False)
(False, False)



### 第三问（10分）：统计与分析——回答小明的问题

基于第二问生成的 `city_data`，定义一个函数 `analyze(data)` 来分析体检结果。该函数需要完成以下工作：

1. 使用**列表推导式**，按下表将所有人分成 A、B、C、D 四组，并打印每组人数：

   | 组别 | 实际状况 | 检测结果 | 通俗说法 |
   |------|---------|---------|---------|
   | A    | 患病    | 阳性    | 查出来了（正确）|
   | B    | 健康    | 阳性    | 虚惊一场（误报）|
   | C    | 健康    | 阴性    | 没事（正确）  |
   | D    | 患病    | 阴性    | 漏掉了（遗漏）|

2. 计算**核心指标**：在所有检测结果为阳性的人当中，真正患病的人占多少比例？打印然后返回该比例。
3. 用文字回答：这个比例与你最初的直觉相符吗？观察 A 组和 B 组的人数，你能解释为什么吗？

In [56]:
# 你的作答
from random import random

def simulate_one_person():
    probability = random()

    if probability < 0.01: # 患病
        ill_res = True
        test_pro = random()
        if test_pro < 0.99:
            test_res = True
        else:
            test_res = False
    else: # 健康
        ill_res = False
        test_pro = random()
        if test_pro < 0.05:
            test_res = True
        else:
            test_res = False

    result_tuple = (ill_res,test_res)
    return result_tuple

city_data = [simulate_one_person() for i in range(100000)]


def analyze(data):

    A = 0
    B = 0
    C = 0
    D = 0 
    
    for people in city_data:
        if people[0] == True and people[1] == True: # 查出来了
            A +=1
        elif people[0] == False and people[1] == True: # 虚惊一场
            B +=1
        elif people[0] == False and people[1] == False: # 没事
            C +=1
        elif people[0] == True and people[1] == False: # 误报
            D +=1

    print("A组人数为：",A)
    print("B组人数为：",B)
    print("C组人数为：",C)
    print("D组人数为：",D)

    positive_people = A + B
    accuracy_rate = A / positive_people

    print("在所有检测结果为阳性的人当中，真正患病的人占的比例为：",accuracy_rate)

    return accuracy_rate

analyze(city_data)


A组人数为： 1025
B组人数为： 4982
C组人数为： 93980
D组人数为： 13
在所有检测结果为阳性的人当中，真正患病的人占的比例为： 0.17063426002996504


0.17063426002996504

你的文字回答
完全不相符
因为患病率极低以及总人口多，所以虽然健康人群的误报率只有5%，但因为健康人数远多于患病人数，所以误报的人数反而超过了正确检出的人数。


### 第四问（10分）：探索不同场景

将第二问和第三问的逻辑合并为一个通用函数：

```python
def run_experiment(n, sick_rate, detect_rate, false_alarm_rate):
    """
    模拟 n 个人的体检，返回"阳性者中真正患病的比例"

    参数：
        n                : 模拟的人数
        sick_rate        : 每个人患病的可能性（如 0.01 表示 1%）
        detect_rate      : 患病者被检出阳性的可能性（如 0.99）
        false_alarm_rate : 健康者被误报为阳性的可能性（如 0.05）
    返回：
        阳性者中真正患病的比例
    """
    # 请完成代码
```

然后调用该函数，探索以下四个场景并记录结果：

```python
# 场景 1：原始设定
print(run_experiment(100000, 0.01, 0.99, 0.05))

# 场景 2：如果患病的人更多（每 100 人中有 10 人）
print(run_experiment(100000, 0.10, 0.99, 0.05))

# 场景 3：如果试剂更精准（误报可能性降至 0.1%）
print(run_experiment(100000, 0.01, 0.99, 0.001))

# 场景 4：如果试剂很差（误报可能性高达 20%）
print(run_experiment(100000, 0.01, 0.99, 0.20))
```

根据实验结果，用文字回答：

1. 对比场景 1 和场景 2：患病人数的多少如何影响了结果？为什么？
2. 对比场景 1 和场景 3：试剂误报可能性的改善，为何对结果影响这么大？
3. 如果你是卫生部门的决策者，你会建议怎样的检测策略？（提示：是否应对阳性者做第二次检测？）

In [66]:
# 你的作答

from random import random

def run_experiment(n, sick_rate, detect_rate, false_alarm_rate):
    
    positive_people = 0
    sick_in_detect = 0

    for i in range(n):
        probability = random()
   
        if probability < sick_rate: # 患病
            ill_res = True
            test_pro = random()
            if test_pro < detect_rate: # 患病且阳性 查出来了
                test_res = True
                positive_people += 1
                sick_in_detect += 1
            else: # 患病却阴性 误报
                test_res = False
        else: # 健康
            ill_res = False
            test_pro = random()
            if test_pro < false_alarm_rate: # 健康却阳性 虚惊一场
                test_res = True
                positive_people += 1
            else:# 健康且阴性 没事
                test_res = False
  
    sick_in_detect_rate = sick_in_detect / positive_people
    return sick_in_detect_rate

# 场景 1：原始设定
print(run_experiment(100000, 0.01, 0.99, 0.05))

# 场景 2：如果患病的人更多（每 100 人中有 10 人）
print(run_experiment(100000, 0.10, 0.99, 0.05))

# 场景 3：如果试剂更精准（误报可能性降至 0.1%）
print(run_experiment(100000, 0.01, 0.99, 0.001))

# 场景 4：如果试剂很差（误报可能性高达 20%）
print(run_experiment(100000, 0.01, 0.99, 0.20))

0.16858813700918965
0.6840440165061898
0.8912213740458015
0.04781473602784761


文字回答
1、患病者更多，阳性者中真正患病的比例就更高。因为当患病率提高时，阳性结果中真阳性的比例会显著增加
2、虽然患病率很低，但假阳性率从5%降到0.1%是50倍的改善。大幅降低假阳性率，能让真阳性在结果中占据主导地位。
3、用成本较低的初筛试剂检测所有人；对初筛阳性者用更精准的试剂复检。


### 总结与思考

通过这次实验，你应该发现了一个反直觉的现象：即使检测试剂"几乎不会漏掉患者"（99% 都能查出来），**一个阳性结果对应的真正患病可能性却可能很低**。这是因为当患病的人本来就很少时，庞大的健康人群中哪怕只有一小部分被误报，其绝对人数也会远超真正的患病者。

在实验中，你实际上反复在思考两个**方向不同**的问题：

- **正向**：已知一个人患病，他被检测为阳性的可能性是多少？（题目直接告诉你了：99%）
- **反向**：已知一个人检测为阳性，他真正患病的可能性是多少？（你通过模拟才发现：只有约 17%）

从"正向"到"反向"，看似只是把条件反过来了，数值却发生了巨大的变化。那么，能不能不靠模拟 10 万个人，而是**直接用一个数学公式**，从"正向"的已知信息算出"反向"的答案呢？

这正是概率论中**贝叶斯定理**（Bayes' Theorem）所解决的问题。在概率论课程中，你将学到这个定理的精确表述和推导。届时你会发现，今天用 10 万次模拟才得到的那个数字，用一行公式就能精确算出——而你通过这次实验建立的直觉，将帮助你更深刻地理解那个公式背后的含义。